File: python/railroaded/models/stop_time.py
from __future__ import annotations

from datetime import time
from enum import Enum
from typing import Optional

import seared as s

from .stop_continuity import StopContinuity


class StopType(Enum):
    '''
    An `Enum` describing the type of the stop.
    '''
    SCHEDULED = 0
    NONE = 1
    VIA_PHONE = 2
    VIA_DRIVER = 3


class Timepoint(Enum):
    '''
    An `Enum` indicating if a stop time is exact or approximate.
    '''
    APPROXIMATE = 0
    EXACT = 1


@s.seared
class StopTime(s.Seared):
    '''
    A GTFS dataclass model for records found in `stop_times.txt`. Identifies a 
    transit route.

    Attributes:
        index (int):
            the sequence index of the stop
        dropoff_booking_id (Optional[str]):
            the unique ID of the dropoff booking rule for the stop
        location_id (Optional[str]):
            the unique ID of the GeoJSON location for the stop
        location_group_id (Optional[str]):
            the unique ID of the location group the stop belongs to
        pickup_booking_id (Optional[str]):
            the unique ID of the pickup booking rule for the stop
        stop_id (Optional[str]):
            the unique ID of the serviced stop
        trip_id (str):
            the unique ID of the trip including the stop
        arrival_time (Optional[str]):
            the arrival time at the stop
        departure_time (Optional[str]):
            the departure time from the stop
        dist_traveled (Optional[float]):
            the distance traveled from the first stop until this stop
        dropoff_continuity (Optional[StopContinuity]):
            the `StopContinuity` for dropoffs at the stop
        dropoff_type (Optional[StopType]):
            the `StopType` for dropoffs at the stop
        end_time (time):
            the end time of the stop
        end_offset (bool):
            a `bool` indicating if the stop date should be offset
        end_pickup_dropoff (Optional[str]):
            the end time for pickup and dropoff
        headsign (Optional[str]):
            the headsign to display when this stop is the destination
        pickup_continuity (Optional[StopContinuity]):
            the `StopContinuity` for pickups at the stop
        pickup_type (Optional[StopType]):
            the `StopType` for pickups at the stop
        start_offset (bool):
            a `bool` indicating if the start date should be offset
        start_pickup_dropoff (Optional[str]):
            the start time for pickup and dropoff
        start_time (time):
            the start time of the stop
        timepoint (Timepoint):
            the timepoint of the stop
    '''

    ### ATTRIBUTES ###
    # Foreign IDs
    dropoff_booking_id: Optional[str] = s.Str(
        data_key='dropoff_booking_rule_id'
    )
    '''the unique ID of the pickup booking rule for the stop'''
    location_id: Optional[str] = s.Str()
    '''the unique ID of the GeoJSON location for the stop'''
    location_group_id: Optional[str] = s.Str()
    '''the unique ID of the location group the stop belongs to'''
    pickup_booking_id: Optional[str] = s.Str(data_key='pickup_booking_rule_id')
    '''the unique ID of the pickup booking rule for the stop'''
    stop_id: Optional[str] = s.Str()
    '''the unique ID of the serviced stop'''
    trip_id: str = s.Str()
    '''the unique ID of the trip including the stop'''

    # Required fields
    index: int = s.Int(data_key='stop_sequence', required=True)
    '''the sequence index of the stop'''
    timepoint: Timepoint = s.Enum(
        enum=Timepoint, missing=Timepoint.EXACT
    )
    '''the timepoint of the stop'''

    # Optional fields
    arrival_time: Optional[str] = s.Str()
    '''the arrival time at the stop'''
    departure_time: Optional[str] = s.Str()
    '''the departure time at the stop'''
    dist_traveled: Optional[float] = s.Float(data_key='shape_dist_traveled')
    '''the distance traveled from the first stop until this stop'''
    dropoff_continuity: Optional[StopContinuity] = s.Enum(
        data_key='continuous_drop_off', enum=StopContinuity
    )
    '''the `StopContinuity` for dropoffs at the stop'''
    dropoff_type: Optional[StopType] = s.Enum(
        data_key='drop_off_type', enum=StopType
    )
    '''the `StopType` for dropoffs at the stop'''
    end_pickup_dropoff: Optional[str] = s.Str(
        data_key='end_pickup_drop_off_window'
    )
    '''the end time for pickup and dropoff'''
    headsign: Optional[str] = s.Str(data_key='stop_headsign')
    '''the headsign to display when this stop is the destination'''
    pickup_continuity: Optional[StopContinuity] = s.Enum(
        data_key='continuous_pickup', enum=StopContinuity
    )
    '''the `StopContinuity` for pickups at the stop'''
    pickup_type: Optional[StopType] = s.Enum(enum=StopType)
    '''the `StopType` for pickups at the stop'''
    start_pickup_dropoff: Optional[str] = s.Str(
        data_key='start_pickup_drop_off_window'
    )
    '''the start time for pickup and dropoff'''


    ### PROPERTIES ###        
    @property
    def end_offset (self) -> bool:
        '''a `bool` indicating if the end date should be offset'''
        return int(self._end_time_str[:2]) >= 24
    
    @property
    def end_time (self) -> time:
        '''the end time of the stop'''
        t = self._end_time_str
        return time(
            hour = 0 if int(t[:2]) >= 24 else int(t[:2]),
            minute = int(t[3:5]),
            seconds = int(t[6:8])
        )
        
    @property
    def _end_time_str (self) -> str:
        '''the end time of the stop as a string'''
        if self.departure_time is None:
            return self.end_pickup_dropoff
        else:
            return self.departure_time
        
    @property
    def start_offset (self) -> bool:
        '''a `bool` indicating if the start date should be offset'''
        return int(self._start_time_str[:2]) >= 24
        
    @property
    def start_time (self) -> time:
        '''the start time of the stop'''
        t = self._start_time_str
        return time(
            hour = 0 if int(t[:2]) >= 24 else int(t[:2]),
            minute = int(t[3:5]),
            seconds = int(t[6:8])
        )
        
    @property
    def _start_time_str (self) -> str:
        '''the start time of the stop as a string'''
        if self.arrival_time is None:
            return self.start_pickup_dropoff
        else:
            return self.arrival_time

File: typescript/src/models/stopTime.ts
import type { StopContinuity, StopType, Time, Timepoint } from '../types'
import {
    fromStopContinuity,
    fromStopType,
    fromTimepoint,
    toStopContinuity,
    toStopType,
    toTimepoint
} from '../types'
import { parseOptionalFloat, parseOptionalInt, parseTime } from '../util'


export type GTFSStopTime = {
    dropoff_booking_rule_id?: string
    location_id?: string
    location_group_id?: string
    pickup_booking_rule_id?: string
    stop_id?: string
    trip_id?: string
    arrival_time?: string
    continuous_drop_off?: string
    continuous_pickup?: string
    departure_time?: string
    drop_off_type?: string
    end_pickup_drop_off_window?: string
    pickup_type?: string
    shape_dist_traveled?: string
    start_pickup_drop_off_window?: string
    stop_headsign?: string
    stop_sequence: string
    timepoint?: string
}


export type MGTFSStopTime = {
    dropoff_booking_id?: string
    location_id?: string
    location_group_id?: string
    pickup_booking_id?: string
    stop_id?: string
    trip_id?: string
    arrival_time?: string
    departure_time?: string
    dist_traveled?: number
    dropoff_continuity?: number
    dropoff_type?: number
    end_pickup_dropoff?: string
    headsign?: string
    index: number
    pickup_continuity?: number
    pickup_type?: number
    start_pickup_dropoff?: string
    timepoint: number
}


export default class StopTime {

    dropoffBookingId?: string
    locationId?: string
    locationGroupId?: string
    pickupBookingId?: string
    stopId?: string
    tripId?: string
    arrivalTime?: string
    departureTime?: string
    distTraveled?: number
    dropoffContinuity?: StopContinuity
    dropoffType?: StopType
    endPickupDropoff?: string
    endTime: Time
    headsign?: string
    index: number
    pickupContinuity?: StopContinuity
    pickupType?: StopType
    startPickupDropoff?: string
    startTime: Time
    timepoint: Timepoint

    constructor (data: MGTFSStopTime) {
        this.dropoffBookingId = data.dropoff_booking_id
        this.locationId = data.location_id
        this.locationGroupId = data.location_group_id
        this.pickupBookingId = data.pickup_booking_id
        this.stopId = data.stop_id
        this.tripId = data.trip_id
        this.arrivalTime = data.arrival_time
        this.departureTime = data.departure_time
        this.distTraveled = data.dist_traveled
        this.dropoffContinuity = data.dropoff_continuity ? 
            toStopContinuity(data.dropoff_continuity) : undefined
        this.dropoffType = data.dropoff_type ?
            toStopType(data.dropoff_type) : undefined
        this.endPickupDropoff = data.end_pickup_dropoff
        this.headsign = data.headsign
        this.index = data.index
        this.pickupContinuity = data.pickup_continuity ?
            toStopContinuity(data.pickup_continuity): undefined
        this.pickupType = data.pickup_type ?
            toStopType(data.pickup_type): undefined
        this.startPickupDropoff = data.start_pickup_dropoff
        this.timepoint = toTimepoint(data.timepoint)

        const endTimeString = this.departureTime ?? this.endPickupDropoff
        const startTimeString = this.arrivalTime ?? this.startPickupDropoff
        if (!endTimeString || !startTimeString) {
            throw new Error(
                `StopTime for stop ${this.stopId ?? 'unknown'} on trip ` +
                `${this.tripId ?? 'unknown'} is missing required time fields`
            )
        }
        this.endTime = parseTime(endTimeString)
        this.startTime = parseTime(startTimeString)
    }

    static fromGTFS (data: GTFSStopTime): MGTFSStopTime {
        return {
            ...data,
            dropoff_booking_id: data.dropoff_booking_rule_id,
            pickup_booking_id: data.pickup_booking_rule_id,
            dist_traveled: parseOptionalFloat(
                data.shape_dist_traveled, 
                undefined
            ),
            dropoff_continuity: parseOptionalInt(
                data.continuous_drop_off,
                undefined
            ),
            dropoff_type: parseOptionalInt(
                data.drop_off_type,
                undefined
            ),
            end_pickup_dropoff: data.end_pickup_drop_off_window,
            headsign: data.stop_headsign,
            index: parseInt(data.stop_sequence),
            pickup_continuity: parseOptionalInt(
                data.continuous_pickup,
                undefined
            ),
            pickup_type: parseOptionalInt(
                data.pickup_type,
                undefined
            ),
            start_pickup_dropoff: data.start_pickup_drop_off_window,
            timepoint: parseOptionalInt(
                data.timepoint,
                fromTimepoint('exact')
            )
        }
    }

    toMGTFS (): MGTFSStopTime {
        return {
            dropoff_booking_id: this.dropoffBookingId,
            location_id: this.locationId,
            location_group_id: this.locationGroupId,
            pickup_booking_id: this.pickupBookingId,
            stop_id: this.stopId,
            trip_id: this.tripId,
            arrival_time: this.arrivalTime,
            departure_time: this.departureTime,
            dist_traveled: this.distTraveled,
            dropoff_continuity: this.dropoffContinuity ?
                fromStopContinuity(this.dropoffContinuity) : undefined,
            dropoff_type: this.dropoffType ?
                fromStopType(this.dropoffType) : undefined,
            end_pickup_dropoff: this.endPickupDropoff,
            headsign: this.headsign,
            index: this.index,
            pickup_continuity: this.pickupContinuity ?
                fromStopContinuity(this.pickupContinuity) : undefined,
            pickup_type: this.pickupType ?
                fromStopType(this.pickupType) : undefined,
            start_pickup_dropoff: this.startPickupDropoff,
            timepoint: fromTimepoint(this.timepoint)
        }
    }

}